Projects/UOContent/Items/Misc/OilFlask.cs

```diff
@@ -0,0 +1,110 @@
+using System.Linq;
+using ModernUO.Serialization;
+using Server.Targeting;
+
+namespace Server.Items;
+
+[SerializationGenerator(0)]
+public partial class OilFlask : Item
+{
+    [Constructible]
+    public OilFlask(int amount = 1) : base(0x1C18)
+    {
+        Stackable = true;
+        Amount = amount;
+    }
+
+    public override int LabelNumber => 1027199; // oil flask
+    public virtual int FilledMessageNumber => 1150864; // You fill the lamp with oil.
+    public virtual double FillMultiplier => 1;
+
+    public override void OnDoubleClick(Mobile from)
+    {
+        from.SendLocalizedMessage(501947); // Select lantern to refuel.
+        from.BeginTarget(2, true, TargetFlags.None, OnTargetPicked);
+    }
+
+    private void OnTargetPicked(Mobile from, object targeted)
+    {
+        if (targeted is Lantern lantern)
+        {
+            if (!lantern.Movable || (lantern.Protected && from.AccessLevel == AccessLevel.Player))
+            {
+                from.SendLocalizedMessage(500685); // You can't use that, it belongs to someone else.
+                return;
+            }
+
+            var wasBurning = lantern.Burning;
+            lantern.Douse();
+            lantern.Duration = Lantern.FullDuration * FillMultiplier;
+            lantern.BurntOut = false;
+
+            if (wasBurning)
+            {
+                lantern.Ignite();
+            }
+
+            from.SendLocalizedMessage(FilledMessageNumber);
+
+            var emptyFlask = new EmptyOilFlask();
+            if (Parent is Container container)
+            {
+                container.AddItem(emptyFlask);
+                emptyFlask.Location = Location;
+            }
+            else
+            {
+                emptyFlask.MoveToWorld(GetWorldLocation(), Map);
+            }
+
+            if (!from.PlaceInBackpack(emptyFlask))
+            {
+                var eable = from.GetItemsInRange(0);
+
+                if (!eable.Any(i => i.StackWith(from, i, false)))
+                {
+                    emptyFlask.MoveToWorld(Location, Map);
+                }
```

Comment:
Let's eliminate LINQ, it has terrible performance.
And I think StackWith should reference `this` as the argument, right?

```suggestion
                foreach (var i in eable)
                {
                    if (i.StackWith(from, this, false))
                    {
                        emptyFlask.MoveToWorld(Location, Map);
                        break;
                    }
                }
```